library/include/DataIdentity.h

```diff
@@ -190,6 +192,102 @@ namespace DFHack
         virtual bool get_item(void *ptr, int idx) = 0;
         virtual void set_item(void *ptr, int idx, bool val) = 0;
     };
+
+    template <typename C>
+    concept isIndexedContainer = requires (C c, C::size_type sz) {
+        { c[sz] } -> std::same_as<typename C::reference>;
+        { c.size() } ->std::same_as<typename C::size_type>;
+        { c.begin() } -> std::same_as<typename C::iterator>;
+        { c.end() } -> std::same_as<typename C::iterator>;
+    };
+
+    template <typename C>
+    concept isResizableContainer = isIndexedContainer<C> && requires (C c, C::iterator i, C::size_type sz, C::value_type v) {
+        { c.erase(i) };
+        { c.resize(sz) };
+        { c.insert(i, v) };
+    };
+
+    template<isIndexedContainer C>
+    class generic_container_identity : public container_identity {
+    private:
+        using T = C::value_type;
+
+        template<typename T> struct type_name {
+            static const inline std::string name = "container";
+        };
+        template<typename T> struct type_name<std::vector<T>> {
+            static const inline std::string name = "vector";
+        };
+        template<typename T> struct type_name<std::deque<T>> {
+            static const inline std::string name = "deque";
+        };
+        template<typename T, int sz> struct type_name<std::array<T, sz>> {
```

Comment:
> I am debating removing the `name = "container"` from the base template to catch this sort of thing.

i was ambivalent on that issue originally. it's not hard for us to add a new specialization but it's more of a burden for a developer of an external plugin. maybe only include the generic case in the library build? that would cause any unspecialized in core code to error at compile time, while allowing plugin developers freedom to use specialized containers for their own purposes (although i'm not sure what the scenario for that is, tbh)